validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java

```diff
@@ -0,0 +1,284 @@
+        }
+        slotToBitlists.put(slot, bitlistToInsert);
+      }
+    }
+
+    for (Attestation sentAttestation : producedAttestations) {
+      Bytes32 sentAttestationDataHash = sentAttestation.getData().hash_tree_root();
+      UInt64 sentAttestationSlot = sentAttestation.getData().getSlot();
+      if (!slotAndBitlistsByAttestationDataHash.containsKey(sentAttestationDataHash)) {
+        continue;
+      }
+      NavigableMap<UInt64, Bitlist> slotAndBitlists =
+          slotAndBitlistsByAttestationDataHash.get(sentAttestationDataHash);
+      for (UInt64 slot : slotAndBitlists.keySet()) {
+        if (slotAndBitlists.get(slot).isSuperSetOf(sentAttestation.getAggregation_bits())) {
+          inclusionDistances.add(slot.minus(sentAttestationSlot).intValue());
+          break;
+        }
+      }
+
+      // Check if the attestation had correct target
+      Bytes32 attestationTargetRoot = sentAttestation.getData().getTarget().getRoot();
+      if (attestationTargetRoot.equals(get_block_root(state, analyzedEpoch))) {
+        correctTargetCount++;
+
+        // Check if the attestation had correct head block root
+        Bytes32 attestationHeadBlockRoot = sentAttestation.getData().getBeacon_block_root();
+        if (attestationHeadBlockRoot.equals(get_block_root_at_slot(state, sentAttestationSlot))) {
+          correctHeadBlockCount++;
+        }
+      }
+    }
+
+    IntSummaryStatistics inclusionDistanceStatistics =
+        inclusionDistances.stream().collect(Collectors.summarizingInt(Integer::intValue));
+
+    // IntSummaryStatistics returns Integer.MIN and MAX when the summarized integer list is empty.
+    return producedAttestations.size() > 0
+        ? new AttestationPerformance(
+            producedAttestations.size(),
+            (int) inclusionDistanceStatistics.getCount(),
+            inclusionDistanceStatistics.getMax(),
+            inclusionDistanceStatistics.getMin(),
+            inclusionDistanceStatistics.getAverage(),
+            correctTargetCount,
+            correctHeadBlockCount)
+        : AttestationPerformance.empty();
+  }
+
+  private Set<BeaconBlock> getBlocksInEpochs(UInt64 startEpochInclusive, UInt64 endEpochExclusive) {
+    UInt64 epochStartSlot = compute_start_slot_at_epoch(startEpochInclusive);
+    UInt64 inclusiveEndEpochEndSlot = compute_start_slot_at_epoch(endEpochExclusive).decrement();
+
+    Set<BeaconBlock> blocksInEpoch = new HashSet<>();
+    UInt64 currSlot = inclusiveEndEpochEndSlot;
+    while (currSlot.isGreaterThanOrEqualTo(epochStartSlot)) {
+      // PerformanceTracker should not be running if chain data is not available.
+      BeaconBlock block =
+          combinedChainDataClient
+              .getBlockInEffectAtSlot(currSlot)
+              .join()
+              .orElseThrow()
+              .getMessage();
+      blocksInEpoch.add(block);
+      if (currSlot.equals(UInt64.ZERO)) {
+        break;
+      }
+      currSlot = currSlot.decrement();
```

Comment:
I think we're now potentially processing blocks multiple times if there are skipped slots.  We need to update the slot based on the returned block's slot:
```suggestion
      if (block.getSlot().equals(UInt64.ZERO)) {
        break;
      }
      currSlot = block.getSlot().decrement();
```

We should add a test that makes sure our calculations are correct when there are skipped slots.

Reply:
Fixed. The current tests already test scenarios with skipped slots.